sql/privilege_test.go

```diff
@@ -90,60 +91,124 @@ func TestPrivilege(t *testing.T) {
-	if err := descriptor.Validate(false); err != nil {
+	descriptor.Grant(security.RootUser, privilege.List{privilege.SELECT})
+	if err := descriptor.Validate(id); err != nil {
 		t.Fatal(err)
 	}
-	descriptor.Revoke("root", privilege.List{privilege.SELECT})
-	if err := descriptor.Validate(false); err == nil {
+	descriptor.Revoke(security.RootUser, privilege.List{privilege.SELECT})
+	if err := descriptor.Validate(id); err == nil {
 		t.Fatal("unexpected success")
 	}
 	// TODO(marc): validate fails here because we do not aggregate
 	// privileges into ALL when all are set.
-	descriptor.Grant("root", privilege.List{privilege.SELECT})
-	if err := descriptor.Validate(false); err == nil {
+	descriptor.Grant(security.RootUser, privilege.List{privilege.SELECT})
+	if err := descriptor.Validate(id); err == nil {
 		t.Fatal("unexpected success")
 	}
-	descriptor.Revoke("root", privilege.List{privilege.ALL})
-	if err := descriptor.Validate(false); err == nil {
+	descriptor.Revoke(security.RootUser, privilege.List{privilege.ALL})
+	if err := descriptor.Validate(id); err == nil {
 		t.Fatal("unexpected success")
 	}
 }
 
+// TestSystemPrivilegeValidate exercises validation for system descriptors.
+// We use 1 (the system database ID).
 func TestSystemPrivilegeValidate(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	descriptor := sql.NewSystemObjectPrivilegeDescriptor()
-	if err := descriptor.Validate(true); err != nil {
-		t.Fatal(err)
-	}
-	descriptor.Grant("foo", privilege.List{privilege.SELECT, privilege.GRANT})
-	if err := descriptor.Validate(true); err != nil {
-		t.Fatal(err)
+	var id sql.ID
+	id = 1
+	allowedPrivileges := sql.SystemAllowedPrivileges[id]
+
+	hasPrivilege := func(pl privilege.List, p privilege.Kind) bool {
+		for _, i := range pl {
+			if i == p {
+				return true
+			}
+		}
+		return false
 	}
-	descriptor.Grant("root", privilege.List{privilege.SELECT})
-	if err := descriptor.Validate(true); err != nil {
+
+	// Exhaustively grant/revoke all privileges.
+	// Due to the way validation is done after Grant/Revoke,
+	// we need to revert the just-performed change after errors.
+	descriptor := sql.NewPrivilegeDescriptor(security.RootUser, allowedPrivileges)
+	if err := descriptor.Validate(id); err != nil {
 		t.Fatal(err)
 	}
-	descriptor.Revoke("root", privilege.List{privilege.SELECT})
-	if err := descriptor.Validate(true); err == nil {
-		t.Fatal("unexpected success")
-	}
-	descriptor.Grant("root", privilege.List{privilege.INSERT})
-	if err := descriptor.Validate(true); err == nil {
-		t.Fatal("unexpected success")
-	}
-	descriptor.Revoke("root", privilege.List{privilege.ALL})
-	if err := descriptor.Validate(true); err == nil {
-		t.Fatal("unexpected success")
+	for _, p := range privilege.ByValue {
+		if hasPrivilege(allowedPrivileges, p) {
+			// Grant allowed privileges. Either they are already
+			// on (noop), or they're accepted.
+			descriptor.Grant(security.RootUser, privilege.List{p})
```

Comment:
even with the comment at L144?
The problem is that validate is after the fact. It would be nice to have grant|revoke just perform their own validation, but that's still a TODO.